addons/folo/common/src/main/java/org/commonjava/indy/folo/change/FoloTrackingAdjustListener.java

```diff
@@ -0,0 +1,128 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.commonjava.indy.folo.change;
+
+import org.commonjava.indy.data.StoreDataManager;
+import org.commonjava.indy.folo.conf.FoloConfig;
+import org.commonjava.indy.folo.data.FoloRecordCache;
+import org.commonjava.indy.folo.model.TrackedContent;
+import org.commonjava.indy.folo.model.TrackedContentEntry;
+import org.commonjava.indy.folo.model.TrackingKey;
+import org.commonjava.indy.model.core.StoreKey;
+import org.commonjava.indy.promote.change.PromoteChangeManager;
+import org.commonjava.indy.promote.change.event.PathsPromoteCompleteEvent;
+import org.commonjava.indy.promote.change.event.PromoteCompleteEvent;
+import org.commonjava.indy.promote.model.PathsPromoteRequest;
+import org.commonjava.indy.promote.model.PathsPromoteResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.event.Observes;
+import javax.inject.Inject;
+
+import java.util.Set;
+
+@ApplicationScoped
+public class FoloTrackingAdjustListener
+{
+
+    private final Logger logger = LoggerFactory.getLogger( getClass() );
+
+    @Inject
+    private FoloConfig foloConfig;
+
+    @Inject
+    private FoloRecordCache recordManager;
+
+    @Inject
+    private PromoteChangeManager promoteChangeManager;
+
+    @Inject
+    private StoreDataManager storeManager;
+
+    public void onPromoteComplete( @Observes final PromoteCompleteEvent event )
+    {
+        logger.trace( "Promote COMPLETE: {}", event );
+
+        if ( event instanceof PathsPromoteCompleteEvent )
+        {
+            handlePathsPromoteComplete( ( (PathsPromoteCompleteEvent) event ).getPromoteResult() );
+        }
+    }
+
+    private void handlePathsPromoteComplete( PathsPromoteResult promoteResult )
+    {
+        PathsPromoteRequest req = promoteResult.getRequest();
+        StoreKey source = req.getSource();
+        StoreKey target = req.getTarget();
+
+        TrackingKey trackingKey = getTrackingKey( source );
+        if ( trackingKey == null )
+        {
+            logger.trace( "No tracking key for access to: {}", source );
+            return;
+        }
+
+        // Get the sealed record, client MUST seal the record before promote
+        TrackedContent trackedContent = recordManager.get( trackingKey );
+        if ( trackedContent == null )
+        {
+            logger.trace( "No sealed record found, trackingKey: {}", trackingKey );
+            return;
+        }
+
+        String error = promoteResult.getError();
+        if ( error != null )
+        {
+            logger.trace( "Error in promoteResult, skip adjust" );
+            return;
+        }
+
+        Set<String> paths = promoteResult.getCompletedPaths();
+
+        if ( paths.isEmpty() )
```

Comment:
I'd move this up to the top of the method to prevent further computation when there are no paths that were promoted.

Reply:
ok